Alina_Kursekova_JavaRush_Project_3

akursekova opened the pull request:
Project 3 - Quest
This is the text game. Each new question of the quest depends on the answer selected before. 

Before to start, please read README.md

akursekova committed:
quest app implementation version 01

akursekova committed:
quest app implementation version 02:
- adding statistics to all jsp with questions
- adding Restart button to each jsp where user lose the game

andriilatyshev commented on src/main/java/dev/akursekova/app/FirstQuestion.java:
```diff
@@ -0,0 +1,33 @@
+            response.sendRedirect("rejectChallenge.jsp");
+        }
+        } else {
+            out.println("<script type=\"text/javascript\">");
```
не хорошо так делать. У тебя по факту кривой хтмл отправляется в браузер. Сделай отдельную jsp для этого.

andriilatyshev commented on src/main/java/dev/akursekova/app/FirstQuestion.java:
```diff
@@ -0,0 +1,33 @@
+        PrintWriter out = response.getWriter();
+
+        if (decision != null) {
+            if (decision.equals("accept")) {
```
просто самое ужасное форматирование, которое только можно придумтаь, никогда так не делай!!! все отступы должны быть соблюдены, а то голову сломать можно.

andriilatyshev commented on src/main/java/dev/akursekova/app/SecondQuestion.java:
```diff
@@ -0,0 +1,34 @@
+
+        if (decision != null) {
+            if (decision.equals("accept")) {
+                response.sendRedirect("thirdQuestion.jsp");
```
если завтра у нас добавится еще 10 вопросов - будем еще десять страниц менять и переписывать тут все? Так дела не делаются. У нас должна быть структура данных, которая хранит все вопросы и варианты ответов. Переходя по ответу, там должна быть ссылка на следующий вопрос.

andriilatyshev replied:
и мы делаем редирект на след вопрос.

akursekova committed:
quest app implementation version 03:
structure of questions was changed: now all questions displayed in only one jsp

akursekova committed:
quest app implementation version 04:
unused jsp removed
added bootstrap modal for statistics
added bootstrap radio buttons

AndryLat commented on src/main/java/dev/akursekova/app/AppServletContextListener.java:
```diff
@@ -0,0 +1,21 @@
+public class AppServletContextListener implements ServletContextListener {
+    @Override
+    public void contextInitialized(ServletContextEvent event) {
+        System.out.println("ServletContextListener started");
```
фу, что это? где логгер?

AndryLat commented on src/main/java/dev/akursekova/app/ApplicationServlet.java:
```diff
@@ -0,0 +1,135 @@
+package dev.akursekova.app;
```
все в одном пакете? Это мы учили все эти модули?

akursekova replied:
Добавила пакет questService  и положила туда классы Question, Answer, QuestService

AndryLat commented on src/main/java/dev/akursekova/app/ApplicationServlet.java:
```diff
@@ -0,0 +1,135 @@
+        System.out.println("doGet started");
+
+        ServletContext context = getServletContext();
+        QuestService questService = (QuestService) context.getAttribute("questService");
```
зачем каждый раз при каждом запросе тянуть этот квест сервис?

akursekova replied:
исправлено
поле вынесла в класс
инициализацию в init метод

AndryLat commented on src/main/java/dev/akursekova/app/ApplicationServlet.java:
```diff
@@ -0,0 +1,135 @@
+        if (session.getAttribute("user") == null){
+            System.out.println("DOGET: this is the beginning of the quest");
+            request.getRequestDispatcher("/index.jsp").forward(request, response);
+        }
```
но код ведь дальше идет при такой логике, так не должно быть.

akursekova replied:
добавила return

AndryLat commented on src/main/java/dev/akursekova/app/ApplicationServlet.java:
```diff
@@ -0,0 +1,135 @@
+        return questRestarted? true : false;
+    }
+
+    private void initializeStatistics(String user, String ipAddress, int numberOfGames, Boolean questRestarted, HttpSession session){
```
я бы в этот метод вынес получение этих параметров тоже.

akursekova replied:
вынесла

AndryLat commented on src/main/java/dev/akursekova/app/ApplicationServlet.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    private boolean gameFinished(Question question){
+        return (question.isWin == true || question.isLoose == true)? true : false;
```
question.isWin == true - пожалуйста, никогда так не делай. 
? true : false; - и вот так тоже.
return (question.isWin || question.isLoose ) - вот как должно выглядеть.

akursekova replied:
все методы с подобным выводом исправлены

AndryLat commented on src/main/java/dev/akursekova/app/QuestService.java:
```diff
@@ -0,0 +1,78 @@
+
+        private static QuestService questServiceInstance;
+
+        Question q1 = Question.builder()
```
я просил не копипастить. Просил же...

AndryLat commented on src/main/java/dev/akursekova/app/QuestService.java:
```diff
@@ -0,0 +1,78 @@
+                .isLoose(true)
+                .build();
+
+        Question q2 = Question.builder()
```
и, извини, это очень стремная реализация. У тебя по факту  сколько будет вопросов, столько и переменных. не должно быть так. Ты же в линкед листе каждый раз ручками не создаешь звено и не сохраняешь отдельно?

akursekova replied:
реализацию исправила согласно комментариям

akursekova committed:
quest app implementation version 05:
QuestService corrected: added init() method whoch returns first question
Logger added

akursekova committed:
quest app implementation version 05:
auto formatting applied
added readme with instruction how to deploy  war file on Tomcat Docker Container

akursekova committed:
quest app implementation version 06:
changed formatting of the Dockefile

akursekova committed:
quest app implementation version 07:
changed formatting to all files

akursekova committed:
quest app implementation version 07:
changed formatting to all files

AndryLat commented on src/main/java/dev/akursekova/app/ApplicationServlet.java:
```diff
@@ -0,0 +1,128 @@
+import java.io.IOException;
+
+@WebServlet(name = "ApplicationServlet", value = "/quest")
+public class ApplicationServlet extends HttpServlet {
```
Почему это называется аппликейшен серрвлет, если он отвечает за квест?

AndryLat commented on src/main/java/dev/akursekova/app/ApplicationServlet.java:
```diff
@@ -0,0 +1,128 @@
+    private static final Logger LOG = LogManager.getLogger(ApplicationServlet.class);
+
+    private QuestService questService = null;
+    private int numberOfGames = 0;
```
Что этО?! у тебя сервлет - ОДИН на ВСЕХ пользователей системы. Это то, что я говорил про многопоточность - в таком случае количество игр будет одно на всех игроков.

AndryLat commented on src/main/java/dev/akursekova/app/ApplicationServlet.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        LOG.info("doGet started");
```
Что тебе дает эта строчка в логах? Правильно, абсолютно ничего. Ты не знаешь какие параметры тут были, что происходило и т.д.

AndryLat commented on src/main/java/dev/akursekova/app/ApplicationServlet.java:
```diff
@@ -0,0 +1,128 @@
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        LOG.info("doPost started");
+        HttpSession session = request.getSession(true);
+        int numberOfGames;
```
а еще ты путаешь локальные переменные и поля класса. В итоге у тебя тут будет проихсодить черти что. И это всплывет только в многопоточной среде, локально ты это не воспроизведешь.

AndryLat commented on src/main/java/dev/akursekova/app/ApplicationServlet.java:
```diff
@@ -0,0 +1,128 @@
+
+    private boolean questRestarted(HttpSession session) {
+        boolean questRestarted = (Boolean) session.getAttribute("questRestarted");
+        return questRestarted ? true : false;
```
а? ты уверена, что проще нельзя?

akursekova committed:
quest app implementation version 08:
added User, UserRepository classes. number of played games stored in the user object.
User object is stored in the session and UserRepository
log4j2 corrected for logging to rolling file

akursekova committed:
quest app implementation version 09:
QuestServletTest added

akursekova committed:
quest app implementation version 10:
ExceptionHandlerServlet added
UserRepositoryTest added

AndryLat commented on src/main/java/dev/akursekova/app/ExceptionHandlerServlet.java:
```diff
@@ -0,0 +1,59 @@
+        out.write("<html><head><title>Exception/Error Details</title></head><body>");
+        if(statusCode != 500){
+            out.write("<h3>Error Details</h3>");
+            out.write("<strong>Status Code</strong>:"+statusCode+"<br>");
```
А почему не перенаправить на jsp? Зачем копировать заранее ужасное решение, к которому ты ранее нигде в коде не прибегала? Копируешь код, улучши его)

AndryLat commented on src/main/java/dev/akursekova/app/QuestServlet.java:
```diff
@@ -0,0 +1,142 @@
+        LOG.info("doGet started. User = " + user + ". Session id = " + session.getId());
+
+        if (!userNameSpecified(session)) {
+            System.out.println("doGet. User name mot specified");
```
есть же логгер, зачем это?

akursekova replied:
Убрала

AndryLat commented on src/test/java/dev/akursekova/app/QuestServletTest.java:
```diff
@@ -0,0 +1,224 @@
+    void doGet_WhenUserNameNotSpecified() throws ServletException, IOException {
+        Mockito.when(request.getAttribute("user")).thenReturn(null);
+        questServlet.doGet(request, response);
+        verify(requestDispatcher).forward(request, response);
```
нет проверки куда перенаправляет...

AndryLat commented on src/test/java/dev/akursekova/app/QuestServletTest.java:
```diff
@@ -0,0 +1,224 @@
+        questServlet.doPost(request, response);
+
+        verify(userRepository, times(1)).add(argThat((user -> {
+           boolean nameMatches = user.getName().equals("test");
```
можно заменить на 
user -> user.getIpAddress().equals("0:0:0:0:0:0:0:1") &&
            user.getNumberOfGames().equals(0) &&
            user.getQuestRestarted().equals(false) )

akursekova replied:
Спасибо! Поправила

akursekova committed:
quest app implementation version 11:
Corrected modal in question.jsp: now the modal fits specified size and no need to scroll it. Footer removed, since it's not needed anymore.
Corrected color of the button which opens modal
Deleted System.output.println
Corrected doGet_WhenUserNameNotSpecified test: added requestDispatcher mock

akursekova committed:
quest app implementation version 12:
Corrected ExceptionHandlerServlet: all html code were substituted by exception.jsp
Added styles_exception.css used by exception.jsp
Added v1 of Readme
Applied autoformatting

akursekova committed:
quest app implementation version 13:
Added v2 of Readme

akursekova committed:
quest app implementation version 13:
Added v2 of Readme

akursekova committed:
quest app implementation version 13:
Added v3 of Readme

akursekova committed:
quest app implementation version 14:
Added v4 of Readme
minor changes related with formating done

akupreychik commented on README.md:
```diff
@@ -1 +1,142 @@
-# javarush-quest
\ No newline at end of file
+# javarush-quest
```
отличное readme

akupreychik commented on src/main/java/dev/akursekova/quest/QuestServlet.java:
```diff
@@ -0,0 +1,134 @@
+            initializeNewUser(session, request);
+            response.sendRedirect(request.getContextPath() + "/quest");
+        } else {
+            String buttonClicked = request.getParameter("answer");
+            if (buttonClicked.equals("Answer")) {
```
какой шанс что тут ты запутаешься? выноси это все в константы с хорошими именами!

akupreychik commented on src/main/java/dev/akursekova/quest/repository/UserRepository.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.HashMap;
+
+public class UserRepository {
+    HashMap<String, User> users = new HashMap<>();
```
модификатор?

akupreychik reviewed (approved):
Супер! Обязательно сделай деплоя на render!

AndryLat commented on pom.xml:
```diff
@@ -0,0 +1,106 @@
+    <packaging>war</packaging>
+    <version>1.0-SNAPSHOT</version>
+    <name>javarush-quest Maven Webapp</name>
+    <url>http://maven.apache.org</url>
```
Тут должна быть ссылка на задеплоенное приложение%)))

AndryLat commented on src/main/java/dev/akursekova/quest/AppServletContextListener.java:
```diff
@@ -0,0 +1,29 @@
+        LOG.info("ServletContextListener started");
+        ServletContext context = event.getServletContext();
+        QuestionService questionService = QuestionService.getInstance();
+        LOG.info("questService has been created: " + questionService);
```
немного свободного пространства между логическими блоками кода не помешает. Я к тому, что надо добавить пустых строк.

AndryLat commented on src/main/java/dev/akursekova/quest/ExceptionHandlerServlet.java:
```diff
@@ -0,0 +1,23 @@
+import java.io.IOException;
+
+@WebServlet(name = "ExceptionHandlerServlet", value = "/exception_handler")
+public class ExceptionHandlerServlet extends HttpServlet {
```
красота! (без сарказма, мне нравится, что ты все вынесла в jsp)

AndryLat commented on src/main/java/dev/akursekova/quest/QuestServlet.java:
```diff
@@ -0,0 +1,134 @@
+        if (!userNameSpecified(session)) {
+            initializeNewUser(session, request);
+            response.sendRedirect(request.getContextPath() + "/quest");
+        } else {
```
Метод сильно сложный. Рзабей на меньшие куски.

AndryLat commented on src/main/java/dev/akursekova/quest/repository/UserRepository.java:
```diff
@@ -0,0 +1,20 @@
+    HashMap<String, User> users = new HashMap<>();
+
+    public void add(User user) {
+        if (user.getName().equals("")) {
```
А если будет состоять только из пробелов? А если будет нал?

akursekova committed:
added maven-surefire-plugin to pom.xml
